doc/coding_guide/Coding_guidelines.md

```diff
@@ -0,0 +1,650 @@
+
+7.  Do not overuse conditional compilation.
+
+-   Do not leave dead code under ```#if 0```. Use comments to hide it from compiler (do not forget to verbose your intention in related comment).
+
+8.  Use macro when extend public API / functionality.
+
+-   Fix binary layout of newly defined structs with ```MFX_PACK_BEGIN_USUAL_STRUCT``` / ```MFX_PACK_BEGIN_STRUCT_W_PTR``` macro (they are expanded to proper ```#pragma pack``` directives) to guarantee compatibility with different compilers.
+
+-   Cover new features which are not in publicly released API yet with API guard. Correct way to protect some feature is to define separate ```#define``` statement under proper ```MFX_VERSION``` guard. Do not use ```MFX_VERSION``` directly in code for such purposes.
+
+-   Prefer positive condition to negative one - ```MFX_ENABLE_FEATURE``` to ```MFX_DISABLE_FEATURE```.
+
+**Types**
+
+9.  Avoid implicit types conversions
+
+-   Declare single parameter constructors as explicit to forbid implicit type conversion.
+
+-   ```std::max```/```min``` won't compile if you mix different types and this is good.
+
+-   Scoped enums will prevent errors of mixing up different enums
+
+-   Be sure to know literal types: 1 - is ```int```, ```NULL``` is ```int```, 1.0 - is ```double```, etc.
+
+-   Use either mfx aliases for types or built in exact-width types (this one is preferred (currently for integer only: ```uint32_t```, ```int32_t```, etc.)). ```int``` is ```mfxI32```, ```double``` is ```mfxF64```, char is ```mfxI8```, etc.
+
+-   Use ```nullptr``` for invalid pointers, not ```NULL```.
+
+10. Use unsigned types whenever it possible. Unsigned overflow is well defined behavior while signed is not.
+
+**Memory management**
+
+11. Avoid manual memory management
+
+-   Functions ```memcpy```, ```memset```, ```sizeof``` are dangerous, use them only if you really need them.
+
+-   Do not copy User Defined Types (UDT) with ```memcpy```, ```operator =``` does the same
+
+-   Do not ```memset``` POD structs for initialization, use ```()``` or ```{}``` initializers.
+
+    Example:
+
+    Instead
+```
+    Class()
+    {
+        memset(&field, 0 , sizeof(field));
+    }
+```
+    Use
+```
+    Class()
+    : field()
+    {}
+```
+-   Do not use ```memcmp``` for comparison of UDT. Even for PODs compiler may fill padding between fields with random garbage which will lead to incorrect comparison result.
+
+-   ```memset```/```memcopy``` non-POD structs will lead to unpredictable fails.
+
+-   Do not use ```memcpy``` if pointers may overlap, use ```memove``` instead
+
+-   ```std::copy``` is better to use for arrays copying. It is safer, it could be faster, because it could benefit from knowing data block size, ```memcpy``` operates with typeless ```void*```. And it works correctly with overlapping blocks (mind the polarity of overlap, maybe you will need ```std::copy_backward```).
```

Comment:
it is not clear that we recommend std::copy/fill whenever possible. It lost in the middle of the section.

Reply:
will rephrase